### ISSUE TITLE
GeeksBLaBla stats

### ISSUE THREAD

**Issue by aboullaite** [geeksblabla/geeksblabla.community#137]
This is more of a discussion. Is there a way where we can get metrics and stats about viewers/listeners? How many hours people have watched/listened ?

**Comment by yjose** [comment]
Yeah, I think it's possible using Facebook graph API https://developers.facebook.com/docs/graph-api/reference/video/video_insights/
but I am not sure if we have access to videos created under the DevC group.

**Comment by yjose** [comment]
also, we can get podcast platforms analytics using anchor account, unfortunately, the only solution is to scrap data using cypress 😀(no open API for anchor)

**Comment by aboullaite** [comment]
You can get videos info from group as well https://developers.facebook.com/docs/graph-api/reference/v7.0/group/videos

**Comment by Kafiil** [comment]
How about google analytics ?

**Comment by aboullaite** [comment]
@Kafiil We want to get number of views and listeners! we need to get them from Facebook, youtube and anchor

**Comment by yjose** [comment]
In Case someone interested to work on this.
The idea is to add a new field for episode Node  https://github.com/DevC-Casa/geeksblabla.com/blob/master/gatsby-node.js#L79
First, you need to create a function `getViewsForVideoId` that returns views number from Facebook API and then you can use it to create a new field for Episode Node called `views`. 

lets me know if you need more info?

**Comment by yjose** [comment]
@ismailElazizi  any idea when we need to add stats ( in the statistics section, the home page ) or for every episode card?

**Comment by zakariaelas** [comment]
@yjose @aboullaite if possible, I would like to work on adding a `views` count for every episode by querying the Facebook Graph API. Also, don't you guys think it would be better to get this data at **runtime on the client** (because views change frequently), as opposed to build time ? I am bringing this up because @yjose mentioned creating a node field for each episode, which from what I know, implies getting data at build time. 
Would love to hear your take on this :pray:

**Comment by yjose** [comment]
Yeah @zakariaelas Go ahead, but I   think calculating views in the build time  is better for page performance, take as example the 'blabla' page, i think it would be very slow as you need to fetch views for 50 episode at once, also fetching views require adding fb sdk to page which will increase load time too.
And we think to add global views to home page in statistics section and this would be hard to calculated in client side.  

Regarding your point of view, i think it would be acceptable to update views every week as we need to rebuild the website on adding new episodes.

**Comment by zakariaelas** [comment]
Thank you @yjose for the clear explanation. At first I thought the views count would appear on the detail view of every episode, which motivated a bit the idea behind realtime fetching. However, I totally agree with what you mentioned, and your approach is better 👌.

**Comment by yjose** [comment]
I would love if you can add it on episode card and episode view too.

**Comment by zakariaelas** [comment]
Upon further inspection, it turns out the FB Graph API does not support stats for videos in facebook groups, but it does for pages. 
From the facebook [docs](https://developers.facebook.com/docs/graph-api/reference/video/video_insights#limitations) under limitations:
> Insights for Videos on Users or Groups are not available

This would imply that some of the episodes would have an empty view count as not all episodes come from the facebook page.

What do you think @yjose ?

**Comment by yjose** [comment]
Hmm , I see 

Maybe we can return an estimation for group videos, as you said this is a limitation from Facebook, so nothing we can do here. 

 I would recommend returning a random number between 2000 - 4000 views in groups Video 😁 , What you think?

**Comment by zakariaelas** [comment]
Sounds good to me. I can't say much on the numbers though because I don't know if Facebook allows you to see the views count on their app as an administrator. I think it's definitely better than leaving other episodes' view count as blank. 

On a side note, one of the **administrators** of the page will need to acquire an **access token** and add it as an **env variable**. This is because facebook only allows admins of a page to query for video insights.
In order to get the facebook token, please read this [guide](https://developers.facebook.com/docs/pages/access-tokens/#access-tokens) from the fb docs. These are the relevant steps summarized below:
1. You will need to [register](https://developers.facebook.com/apps) your facebook account on the Facebook Developers platform. 
2. Get a short lived **user** access token from the Graph Explorer tool.
2. Get a long lived **user** access token: This is particularly useful because it will allow you to get a permanent token, meaning you will not have to acquire a new one everytime you build (short lived access tokens expire in an hour for facebook pages).
3. Get a **page** access token using the long lived user token from step 2 above. This token is what will allow the app to access information about videos published on the page. 

**Why do we need a user access token AND a page access token?**

Just in case you are wondering just like I did, this answered it for me:
> To get a Page access token, you will need a User access token. This User access token must be generated by a person who can perform the action on the Page.

**Comment by yjose** [comment]
thank you @zakariaelas for the great explanation, will generate tokens and back to you

**Comment by zakariaelas** [comment]
@yjose you're welcome :+1: .

I played a bit with the positioning of the views count. Initially, I thought about having an "eye" icon just to save up some space, but I'm not sure how you guys would get the svg for that.

Anyways, please let me know what you think. Would love to hear other suggestions :smiley: .

![image](https://user-images.githubusercontent.com/33696020/88488033-755dc880-cf82-11ea-8113-02bbcdf9847c.png)

**Comment by yjose** [comment]
I think your solution is clean. Go gor it, i would recommend using x.x k format ( 2.1k) for episode with views more than 1k. 

For the card. I am ok with it 
For episode view. I would prefer it under date 🤔

**Comment by yjose** [comment]
@zakariaelas  Unfortunately we need to wait for Facebook app review to make sure the access token will work as expected. this is something new that Facebook has implemented in the last few months due to all the data leaking controversy 😏

**Comment by zakariaelas** [comment]
@yjose thank you for the feedback. 

It's okay, i'll just keep working on my fork until you hear back from facebook 👍👌
